pkg/router/template/configmanager/haproxy/client.go

```diff
@@ -0,0 +1,204 @@
+package haproxy
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"time"
+
+	haproxy "github.com/bcicen/go-haproxy"
+	"github.com/golang/glog"
+)
+
+const (
+	// Prefix for the socket file used for haproxy dynamic API commands.
+	afUnixSocketPrefix = "unix://"
+
+	// Prefix if TCP is used to communicate with haproxy.
+	tcpSocketPrefix = "tcp://"
+
+	// maxRetries is the number of times a command is retried.
+	maxRetries = 3
+)
+
+// Client is a client used to dynamically configure haproxy.
+type Client struct {
+	socketAddress string
+	timeout       int
+
+	backends []*Backend
+	maps     map[string]*HAProxyMap
+}
+
+// NewClient returns a client used to dynamically change the haproxy config.
+func NewClient(socketName string, timeout int) *Client {
+	sockAddr := socketName
+	if !strings.HasPrefix(sockAddr, afUnixSocketPrefix) && !strings.HasPrefix(sockAddr, tcpSocketPrefix) {
+		sockAddr = fmt.Sprintf("%s%s", afUnixSocketPrefix, sockAddr)
+	}
+
+	return &Client{
+		socketAddress: sockAddr,
+		timeout:       timeout,
+		backends:      make([]*Backend, 0),
+		maps:          make(map[string]*HAProxyMap),
+	}
+}
+
+// RunCommand executes a haproxy dynamic config API command and if present
+// converts the response as desired.
+func (c *Client) RunCommand(cmd string, converter Converter) ([]byte, error) {
+	glog.V(4).Infof("Running haproxy command: %q ...", cmd)
+	buffer, err := c.runCommandWithRetries(cmd, maxRetries)
+	if err != nil {
+		glog.Warningf("haproxy dynamic config API command %s failed with error: %v", cmd, err)
+		return nil, err
+	}
+
+	response := buffer.Bytes()
+	glog.V(4).Infof("haproxy command response: %q", string(response))
+	if converter == nil {
+		return response, nil
+	}
+
+	return converter.Convert(response)
+}
+
+// Execute runs a haproxy dynamic config API command.
+func (c *Client) Execute(cmd string) ([]byte, error) {
+	return c.RunCommand(cmd, nil)
+}
+
+// Reset resets any changes and clears the backends and maps.
+func (c *Client) Reset() {
+	c.backends = make([]*Backend, 0)
+	c.maps = make(map[string]*HAProxyMap)
+}
+
+// Commit flushes out any pending changes on all the backends and maps.
+func (c *Client) Commit() error {
+	for _, b := range c.backends {
+		if err := b.Commit(); err != nil {
+			return err
+		}
+	}
+
+	for _, m := range c.maps {
+		if err := m.Commit(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Backends returns the list of configured haproxy backends.
+func (c *Client) Backends() ([]*Backend, error) {
+	if len(c.backends) == 0 {
+		if backends, err := GetHAProxyBackends(c); err != nil {
```

Comment:
If these methods don't need to be public (`GetHAProxyBackends`) they shouldn't be public.